cors.go

```diff
@@ -1,22 +1,113 @@
 package handlers
 
-import "net/http"
+import (
+	"net/http"
+	"strconv"
+	"strings"
+)
 
 // CORSOption represents a functional option for configuring the CORS middleware.
 type CORSOption func(*cors) error
 
 type cors struct {
-	h              http.Handler
-	allowedHeaders []string
-	allowedMethods []string
-	allowedOrigins []string
-	maxAge         int
-	ignoreOptions  bool
+	h                http.Handler
+	allowedHeaders   []string
+	allowedMethods   []string
+	allowedOrigins   []string
+	exposedHeaders   []string
+	maxAge           int
+	ignoreOptions    bool
+	allowCredentials bool
 }
 
+var (
+	defaultCorsMethods = []string{"GET", "HEAD", "POST"}
+	defaultCorsHeaders = []string{"Accept", "Accept-Language", "Content-Language"}
+	nilHandlerFunc     = http.HandlerFunc(nilHandler)
+)
+
+const (
+	corsOptionMethod           string = "OPTIONS"
+	corsAllowOriginHeader      string = "Access-Control-Allow-Origin"
+	corsExposeHeadersHeader    string = "Access-Control-Expose-Headers"
+	corsMaxAgeHeader           string = "Access-Control-Max-Age"
+	corsAllowMethodsHeader     string = "Access-Control-Allow-Methods"
+	corsAllowHeadersHeader     string = "Access-Control-Allow-Headers"
+	corsAllowCredentialsHeader string = "Access-Control-Allow-Credentials"
+	corsRequestMethodHeader    string = "Access-Control-Request-Method"
+	corsRequestHeadersHeader   string = "Access-Control-Request-Headers"
+	corsOriginHeader           string = "Origin"
+	corsVaryHeader             string = "Vary"
+	corsOriginMatchAll         string = "*"
+)
+
 func (ch *cors) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	handler := ch.h
+	defer func() {
+		handler.ServeHTTP(w, r)
+	}()
+
+	if origin := r.Header.Get(corsOriginHeader); ch.isOriginAllowed(origin) {
+		if r.Method == corsOptionMethod {
+			if ch.ignoreOptions {
+				return
+			}
+
+			handler = nilHandlerFunc
+			if _, ok := r.Header[corsRequestMethodHeader]; !ok {
+				w.WriteHeader(http.StatusBadRequest)
```

Comment:
This looks good to me.

You *could* provide a way to provide custom error response handlers later but you either need to provide one for the major cases (bad req, forbidden, methods not allowed) or somehow provide that information to the package user (e.g. `handlers.CORSFailureReason(r)` via the request context.

(I suggest we keep it simple for now and add customisation later. A simple HTTP error status with no body is probably fine for the majority of use cases when in comes to CORS, since the remote end can reformat the error/display a pretty error page as needed)